### PR TITLE
feat: support for problem json

### PR DESCRIPTION
**Pull request by banartdev**
Implements reading an `application/problem+json` from the response of API calls. This data is then used for display in the `ErrorAlert`.

Refs #46 

## Coder

- [x] I paid attention to the warnings and errors the IDE gave me and I have fixed them where appropriate
- [x] I wrote unit/integration tests where applicable
- [x] I tested that the issue is resolved
- [x] I tested that no regressions were introduced (other features still work)
- [ ] I tested edge cases
- [ ] I tested error cases and errors are handled correctly
- [x] The build and tests have succeeded
- [x] There are no conflicts with the main / master branch

## Reviewer

Code review:
- [x] The coder completed their checklist to a reasonable extent (relative to the complexity of the PR)
- [x] I have reviewed the code, and it is easy to understand and follows best practices.

Testing:
- [x] I have tested that the issue is resolved
- [ ] I have tested that no regressions were introduced (other features still work)
- [ ] I have tested edge cases
- [ ] I have tested error cases and errors are handled correctly

Merging:
- [ ] I have resolved any database version conflicts
- [x] The build and tests have succeeded
- [ ] I have moved the issue to *Waiting Deployment*

### CODE REVIEW

File: src/ErrorAlert.tsx [hunk 1]
```diff
@@ -1,36 +1,48 @@
-import { Alert, AlertTitle } from '@mui/material';
+import {Alert, AlertTitle, Typography} from '@mui/material';
 import * as React from "react";
 import {ErrorIcon} from "./icons";
 import {ForwardedRef} from "react";
+import {ApiError, ProblemJson} from "./fetch";
 
-export const ErrorAlert = React.forwardRef(({error}: {error: any}, ref: ForwardedRef<HTMLDivElement>) => {
+export const ErrorAlert = React.forwardRef(({error}: {error: Partial<ProblemJson> | Error | string | any}, ref: ForwardedRef<HTMLDivElement>) => {
```
Comment (rKatex) [inline src/ErrorAlert.tsx:7]: I think `ApiError` is missing from these params. 

And I think removing `any` here would make it more robust and typesafe.

File: src/ErrorAlert.tsx [hunk 2]
```diff
@@ -1,36 +1,48 @@
...
-            message: error.status ? `${error.status} ${error.message}` : error.message,
-            details: error.details,
-        };
-    }
-    else {
-        return {
-            message: error.toString(),
-        };
-    }
+function normalizeError(error: any): Pick<ProblemJson, "title"> & Partial<ProblemJson> & Partial<Error> {
```
Comment (rKatex) [inline src/ErrorAlert.tsx:26]: I think the `any` here could be removed as well in favour of the same type as above.

Also, I would consider changing the return type to a defined. Currently its a bit hard to see what possible fields are there in the return object.
 
```typescript
    type NormalizedError = {
        title: string;
        status?: number;
        detail?: string;
        type?: string;
        instance?: string;
        stack?: string;
        [key: string]: unknown;
    };
```

File: src/ErrorAlert.tsx (hunk 1 above)
Comment (rKatex) [inline src/ErrorAlert.tsx:7]: Maybe an optional param `showStack` would make sense, to avoid showing the stack to a user if unwanted.

File: src/ErrorAlert.tsx [hunk 3]
```diff
@@ -1,36 +1,48 @@
...
 
-export const ErrorAlert = React.forwardRef(({error}: {error: any}, ref: ForwardedRef<HTMLDivElement>) => {
+export const ErrorAlert = React.forwardRef(({error}: {error: Partial<ProblemJson> | Error | string | any}, ref: ForwardedRef<HTMLDivElement>) => {
     if (!error) {
         return null;
     }
-    const {message, details} = formatError(error);
+    const {title, status, detail, type, instance, stack, ...rest} = normalizeError(error);
     return (
         <Alert ref={ref} color="error" icon={<ErrorIcon/>}>
```
Comment (rKatex) [inline src/ErrorAlert.tsx:13]: Not your change, but I think using `severity` for alerts is the preferred way. `color` should only be used if the color of the severity is not ok.

File: src/ErrorAlert.tsx [hunk 4]
```diff
@@ -1,36 +1,48 @@
...
-            <AlertTitle>{message}</AlertTitle>
-            {details}
+            <AlertTitle>{title}</AlertTitle>
+			{detail}
+			{status && <Typography component="div" variant="caption">Status: {status}</Typography>}
+			{instance && <Typography component="div" variant="caption">Instance: {instance}</Typography>}
+			{type && type != "about:blank" && <Typography component="div" variant="caption">Type: {type}</Typography>}
+			{Object.keys(rest).map(key => (
```
Comment (rKatex) [inline src/ErrorAlert.tsx:19]: `Object.entries(rest).map(([key, value]) =>` could be used here to avoid the additional access for the value in the loop

File: src/fetch.ts [hunk 5]
```diff
@@ -42,31 +61,39 @@ export type ApiFetcher = ReturnType<typeof createApiFetcher>;
...
+export async function createApiError(response: Response) {
 	try {
 		const error = await response.json();
-		if (error?.error && error.message) {
-			return {
-				message: error.error as string,
-				details: error.message as string,
-			};
+		// Assume a problem json, if title and status are present and of the correct type
+		if (typeof error.title == "string" && typeof error.status == "number") {
```
Comment (rKatex) [inline src/fetch.ts:72]: Just a minor thing, but extracting the condition into a method e.g. `isProblemJson(error)` could improve the readability. This might result in removing the comment as well, as the method states whats happening already.

**Review by rKatex — APPROVED** [review]
Most of these things are cosmetic but I think could improve the readability. No need to do the changes right away - but still I would recommend it

File: src/fetch.ts [hunk 6]
```diff
@@ -42,31 +61,39 @@ export type ApiFetcher = ReturnType<typeof createApiFetcher>;
...
+				title: error.error as string,
+				status: response.status,
+				detail: error.message as string,
+			});
 		}
+		// As a fallback, just copy the response body to the error and use title and status
+		// from the response
+		return new ApiError({
+			...error,
+			title: response.statusText,
```
Comment (rKatex) [inline src/fetch.ts:87]: Would it make sense, to have `?? 'Unknown Error'` for the title in this case?

File: src/ErrorAlert.tsx (hunk 1 above)
Comment (peterkarrer) [reply src/ErrorAlert.tsx:7]: There are two reasons for `any` here:
 - Backwards compatibility
 - Ease of use. You don't need to know the type of the error caught in a try-catch or returned from `useAsync` in order to display it.
 
`ApiError` is already covered by `Partial<ProblemJson>` and `Error`.

File: src/ErrorAlert.tsx (hunk 2 above)
Comment (peterkarrer) [reply src/ErrorAlert.tsx:26]: I prefer to construct the type from `ProblemJson` and `Error`. However, I will extract the type and add comments.

File: src/ErrorAlert.tsx (hunk 1 above)
Comment (peterkarrer) [reply src/ErrorAlert.tsx:7]: Good point. At the moment the stack is actually not shown at all. I originally planned for that, but showing the stack is almost always too verbose. Devs can just open DevTools to figure it out.

File: src/ErrorAlert.tsx (hunk 3 above)
Comment (peterkarrer) [reply src/ErrorAlert.tsx:13]: I'll change that.

File: src/fetch.ts (hunk 5 above)
Comment (peterkarrer) [reply src/fetch.ts:72]: Agreed

File: src/fetch.ts (hunk 6 above)
Comment (peterkarrer) [reply src/fetch.ts:87]: The `statusText` property is always set, however it can be an empty string (and [always will for HTTP/2 requests](https://fetch.spec.whatwg.org/#concept-response-status-message)). I will handle this case.

**Review by rKatex — APPROVED** [review]
